sql/analyzer/fix_exec_indexes.go

```diff
@@ -389,7 +413,71 @@ func (s *idxScope) finalizeSelf(n sql.Node) (sql.Node, error) {
 		nn.OnDupExprs = s.expressions
 		return nn.WithChecks(s.checks), nil
 	default:
-		// child scopes don't account for projections
+
+		// fill in column ids
+		switch n := n.(type) {
+		case *plan.Project:
+			for _, e := range n.Projections {
+				if ide, ok := e.(sql.IdExpression); ok {
+					s.ids = append(s.ids, ide.Id())
+				} else {
+					s.ids = append(s.ids, 0)
+				}
+			}
+		case *plan.ResolvedTable, *plan.IndexedTableAccess, *plan.SubqueryAlias, *plan.RecursiveTable, *plan.RecursiveCte, *plan.SetOp, *plan.ValueDerivedTable, *plan.JSONTable:
+			if rt, ok := n.(*plan.ResolvedTable); ok && plan.IsDualTable(rt.Table) {
+				s.ids = append(s.ids, 0)
+				break
+			}
+			cols := n.(plan.TableIdNode).Columns()
+			if tn, ok := n.(sql.TableNode); ok {
+				if pkt, ok := tn.UnderlyingTable().(sql.PrimaryKeyTable); ok && len(pkt.PrimaryKeySchema().Schema) != len(n.Schema()) {
+					firstcol, _ := cols.Next(1)
+					for _, c := range n.Schema() {
+						ord := pkt.PrimaryKeySchema().IndexOfColName(c.Name)
+						colId := firstcol + sql.ColumnId(ord)
+						s.ids = append(s.ids, colId)
+					}
+					break
+				}
+			}
+			cols.ForEach(func(col sql.ColumnId) {
+				s.ids = append(s.ids, col)
+			})
+
+		case *plan.Window:
+			for _, e := range n.SelectExprs {
+				if ide, ok := e.(sql.IdExpression); ok {
+					s.ids = append(s.ids, ide.Id())
+				}
+			}
+		case *plan.GroupBy:
+			for _, e := range n.SelectedExprs {
+				if ide, ok := e.(sql.IdExpression); ok {
+					s.ids = append(s.ids, ide.Id())
+				}
+			}
```

Comment:
Are there any interfaces we can use to make sure we're grabbing all the right types and not missing any? Seems tricky to make sure to we have every struct listed and to keep it up to date as we add new ones. For example, could we use `Projector` to collapse `Window`, `GroupBy`, and `Project` into a single case?

Reply:
Projector is a good idea. Covering all types would look something like the memo IR absorbing later half of transformation rules, or sql.Node implementing the relational expressions we track for memo nodes. Need everything to implement a common interface.